### DISCUSSION TITLE
How can I contribute about typing issue and make it support both of version 1.4 and 2+?

### DISCUSSION THREAD

**Discussion by fResult** [sqlalchemy/sqlalchemy#11496] [Usage Questions]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [x] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

How can we make it support back to the SQLAlchemy version 1.4?

As I forked this repository and found the `or_` and `and_` functions, they already have a type

But version 1.4 isn't (my company is not ready to upgrade to v.2)

The `public_factory` function is inferred as `Any` (here is the SQLAlchemy 1.4 lib code on my local)
![image](https://github.com/sqlalchemy/sqlalchemy/assets/19329932/76849b67-64cd-4e3a-86c4-02b370158626)

If I do something like this...
![image](https://github.com/sqlalchemy/sqlalchemy/assets/19329932/f02ba6b1-03f8-4d0b-9b34-e8dc41c93931)

... it's work

![image](https://github.com/sqlalchemy/sqlalchemy/assets/19329932/77750d00-1f52-4475-be95-8e1b66e2823a)

### To Reproduce

```python
Already provided above
```

### Error

```

``` 

### Versions

- OS: MacOS
- Python: 3.10.*
- SQLAlchemy: 1.4.*
- Type checker:  Pylance VSCode extension v.2024.6.1

**Comment by CaselIT** [comment]
Hi,

For 1.4 you have the stub package https://github.com/sqlalchemy/sqlalchemy2-stubs.
But that's best effort support only.

PR will not be accepted to type the 1.4 branch, since 1.4 is in major bug fix only mode currently.

If you care about typing you should upgrade to version 2.0
